core/src/main/java/org/elasticsearch/cluster/routing/UnassignedMeta.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.routing;
+
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.joda.FormatDateTimeFormatter;
+import org.elasticsearch.common.joda.Joda;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+
+/**
+ * Holds additional information as to why the shard is in unassigned state.
+ */
+public class UnassignedMeta implements ToXContent, Writeable<UnassignedMeta> {
+
+    public static final FormatDateTimeFormatter DATE_TIME_FORMATTER = Joda.forPattern("dateOptionalTime");
+
+    /**
+     * Reason why the shard is in unassigned state.
+     * <p/>
+     * Note, ordering of the enum is important, make sure to add new values
+     * at the end and handle version serialization properly.
+     */
+    public enum Reason {
+        /**
+         * Unassigned as a result of an API creation of an index.
+         */
+        NEW,
+        /**
+         * Unassigned as a result of a full cluster recovery.
+         */
+        CLUSTER_RECOVERY,
+        /**
+         * Unassigned as a result of opening a closed index.
+         */
+        OPEN,
+        /**
+         * Unassigned as a result of importing a dangling index.
+         */
+        DANGLING_IMPORT,
+        /**
+         * Unassigned as a result of restoring into a new index.
+         */
+        NEW_RESTORE,
+        /**
+         * Unassigned as a result of restoring into a closed index.
```

Comment:
This one confused me because I didn't realize how restoring into closed indices worked. After checking the code, I think it's good to to add "into a closed index, causing it to be opened".